internal/graph/check_test.go

```diff
@@ -3072,3 +3076,992 @@ func TestConsumeUsersets(t *testing.T) {
 		})
 	}
 }
+
+func TestDispatch(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	checker := NewLocalChecker()
+	defer checker.Close()
+	mockResolver := NewMockCheckResolver(ctrl)
+	checker.SetDelegate(mockResolver)
+
+	parentReq := &ResolveCheckRequest{
+		TupleKey:        tuple.NewTupleKeyWithCondition("document:doc1", "viewer", "user:maria", "condition1", nil),
+		RequestMetadata: NewCheckRequestMetadata(20),
+	}
+	tk := tuple.NewTupleKeyWithCondition("group:1", "member", "user:maria", "condition1", nil)
+
+	expectedReq := &ResolveCheckRequest{
+		TupleKey:        tuple.NewTupleKeyWithCondition("group:1", "member", "user:maria", "condition1", nil),
+		RequestMetadata: NewCheckRequestMetadata(19),
+	}
+
+	var req *ResolveCheckRequest
+	mockResolver.EXPECT().ResolveCheck(gomock.Any(), gomock.AssignableToTypeOf(req)).DoAndReturn(
+		func(_ context.Context, req *ResolveCheckRequest) (*ResolveCheckResponse, error) {
+			require.Equal(t, expectedReq.GetTupleKey(), req.GetTupleKey())
+			require.Equal(t, expectedReq.GetRequestMetadata().Depth, req.GetRequestMetadata().Depth)
+			require.Equal(t, uint32(1), req.GetRequestMetadata().DispatchCounter.Load())
+			return nil, nil
+		})
+	dispatch := checker.dispatch(context.Background(), parentReq, tk)
+	_, _ = dispatch(context.Background())
+}
+
+// helperReceivedMessage is a helper function to listen to dispatchMsg and create
+// a list of dispatchMsg.
+func helperReceivedMessage(dispatchChan chan dispatchMsg) []dispatchMsg {
+	var receivedDispatches []dispatchMsg
+	for msg := range dispatchChan {
+		receivedDispatches = append(receivedDispatches, msg)
+	}
+	return receivedDispatches
+}
+
+// dispatchMsgAssertion is helper structure to allow easy assertion on dispatchMsg.
+type dispatchMsgAssertion struct {
```

Comment:
It's a code smell to me that you had to create this when you already have `dispatchMsg`.

`dispatchMsg` doesn't need to include a CheckHandlerFunc. You can replace it with what the `dispatch` function takes as parameters. And then this code:


```go
pool.Go(func(ctx context.Context) error {
	resp, err := msg.dispatch(ctx)
	outcomes <- checkOutcome{resp: resp, err: err}
	return nil
})
```

can become

```go
pool.Go(func(ctx context.Context) error {
	resp, err := c.dispatch(ctx, msg.parentReq, msg.tupleKey)
	outcomes <- checkOutcome{resp: resp, err: err}
	return nil
})
```

Taking it one step further, we could (not in this PR):


```go
type dispatchParams struct {
	parentReq *ResolveCheckRequest
	tk        *openfgav1.TupleKey
}

type dispatchMsg struct {
	err            error
	shortCircuit   bool
	dispatchParams *dispatchParams
}

func (c *LocalChecker) dispatch(_ context.Context, params dispatchParams) CheckHandlerFunc {
```

Reply:
I will use the dispatchParams as it makes the unit test much easier to construct